src/Functions/FunctionsStringArray.h

```diff
@@ -271,6 +292,42 @@ class SplitByCharImpl
             throw Exception("Illegal separator for function " + getName() + ". Must be exactly one byte.", ErrorCodes::BAD_ARGUMENTS);
 
         sep = sep_str[0];
+
+        if (arguments.size() > 2)
+        {
+            std::optional<Int64> max_split_opt = std::nullopt;
+            if (!((max_split_opt = getMaxSplit<UInt8>(arguments[2]))
+                || (max_split_opt = getMaxSplit<Int8>(arguments[2]))
+                || (max_split_opt = getMaxSplit<UInt16>(arguments[2]))
+                || (max_split_opt = getMaxSplit<Int16>(arguments[2]))
+                || (max_split_opt = getMaxSplit<UInt32>(arguments[2]))
+                || (max_split_opt = getMaxSplit<Int32>(arguments[2]))
+                || (max_split_opt = getMaxSplit<UInt64>(arguments[2]))
+                || (max_split_opt = getMaxSplit<Int64>(arguments[2]))))
+            {
+                throw Exception(
+                    ErrorCodes::ILLEGAL_COLUMN,
+                    "Illegal column {} of third argument of function {}",
+                    arguments[2].column->getName(),
+                    getName());
+            }
+            max_split = *max_split_opt;
+        }
+    }
+
+    template <typename DataType>
+    std::optional<Int64> getMaxSplit(const ColumnWithTypeAndName & argument)
+    {
+        const auto * col = checkAndGetColumnConst<ColumnVector<DataType>>(argument.column.get());
+        if (!col)
+            return std::nullopt;
+
+        Int64 result= static_cast<Int64>(col->template getValue<DataType>());
+        if (result < 0 && result != -1)
```

Comment:
Could you please add comment how this function behave if max_split = -1. It seems that it should not do anything because in check we have `max_split >= 0`.

Reply:
If this function cannot take max_split as negative integer, and argument can be only constant number, we can validate columns in validate function, then to getMaxSplit we can just get field, and from field get UInt64 value.

Reply:
I understand what you mean. Since python function allows max_split to to be negative integer, maybe it is better to keep consistent with it in ClickHouse?

Reply:
It seems that in ClickHouse we already have `splitByChar` and we do not need to pass -1 for default behavior. It is actually inconsistent with other ClickHouse functions, I am not aware if we pass somewhere -1 for default behavior.

Reply:
Default behavior is not changed because when number of split is limited only when max_split >= 0.  See function `SplitByCharImpl::get`

Reply:
`max_split == -1` in python means there are no limit on split, not that we do not need split. 
```
Return a list of the words in the string, using sep as the delimiter string. If maxsplit is given, at most maxsplit splits are done (thus, the list will have at most maxsplit+1 elements). If maxsplit is not specified or -1, then there is no limit on the number of splits (all possible splits are made).
```

Reply:
If we use Int64 for max_split only to hold -1 if there are no split, it can be a problem if we want to have split with unsigned value greater than maximum value that Int64 can hold. It seems that we can just store std::optional<UInt64> and it will be fine.

Reply:
Ok.